Trying to populate optional array on struct results in compile error

### Describe the bug

See code sample. I have a struct, with optional authors field.

### Expected Behavior

Code compiles, allows post.authors to be set

### Current Behavior

Errors with:

```
/tmp/v_501/test.14504799408771845083.tmp.c:2260:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
/tmp/v_501/test.14504799408771845083.tmp.c:12500:29: error: assigning to 'Array_string' (aka 'struct array') from incompatible type '_option_Array_string' (aka 'struct _option_Array_string')
                *(Array_string*) _t2.data = post->authors;
                                          ^ ~~~~~~~~~~~~~
1 warning and 1 error generated.
```

### Reproduction Steps

```
struct Post {
mut:
  authors ?[]string
}

fn populate(mut post Post) []string {
  return post.authors or {
    post.authors = []string{}
    post.authors
  }
}

fn main() {
  mut test := Post {}
  populate(mut test)
  println(test)
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

0.3.3

### Environment details (OS name and version, etc.)

```
V full version: V 0.3.3 fedf0f7
OS: macos, macOS, 13.3, 22E252
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz

getwd: /Users/kieran/Work/gol_implementations/v
vexe: /usr/local/Cellar/vlang/0.2.4/libexec/v
vexe mtime: 2023-04-01 04:45:27

vroot: OK, value: /usr/local/Cellar/vlang/0.2.4/libexec
VMODULES: OK, value: /Users/kieran/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.40.0
Git vroot status: weekly.2023.13-27-gfedf0f79
.git/config present: true

CC version: Apple clang version 14.0.3 (clang-1403.0.22.14.1)
thirdparty/tcc status: thirdparty-macos-amd64 46662e20
```

Another slightly simpler test case using optional string:

```
struct Post {
mut:
  author ?string
}

fn populate(mut post Post) string {
  return post.author or {
    post.author = "Test"
    post.author
  }
}

fn main() {
  mut test := Post {}
  populate(mut test)
  println(test)
}
```

Error:

```
/tmp/v_501/test.9737649463928909276.tmp.c:2258:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
/tmp/v_501/test.9737649463928909276.tmp.c:12498:23: error: assigning to 'string' (aka 'struct string') from incompatible type '_option_string' (aka 'struct _option_string')
                *(string*) _t2.data = post->author;
                                    ^ ~~~~~~~~~~~~
1 warning and 1 error generated.
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1127,8 +1127,7 @@ fn (mut c Checker) check_or_last_stmt(stmt ast.Stmt, ret_type ast.Type, expr_ret
 				c.expected_or_type = ret_type.clear_flags(.option, .result)
 				last_stmt_typ := c.expr(stmt.expr)
 
-				if ret_type.has_flag(.option)
-					&& (last_stmt_typ.has_flag(.option) || last_stmt_typ == ast.none_type) {
+				if last_stmt_typ.has_flag(.option) || last_stmt_typ == ast.none_type {
 					if stmt.expr in [ast.Ident, ast.SelectorExpr, ast.CallExpr, ast.None] {
 						expected_type_name := c.table.type_to_str(ret_type.clear_flags(.option,
 							.result))
@@ -1440,6 +1439,7 @@ fn (mut c Checker) selector_expr(mut node ast.SelectorExpr) ast.Type {
 		if node.or_block.kind == .block {
 			c.expected_or_type = node.typ.clear_flags(.option, .result)
 			c.stmts_ending_with_expression(node.or_block.stmts)
+			c.check_or_expr(node.or_block, node.typ, c.expected_or_type, node)
 			c.expected_or_type = ast.void_type
 		}
 		return field.typ

--- a/vlib/v/checker/tests/option_selector_or_block_err.out
+++ b/vlib/v/checker/tests/option_selector_or_block_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/option_selector_or_block_err.vv:9:10: error: `or` block must provide a value of type `[]string`, not `?[]string`
+    7 |   return post.authors or {
+    8 |     post.authors = []string{}
+    9 |     post.authors
+      |          ~~~~~~~
+   10 |   }
+   11 | }

--- a/vlib/v/checker/tests/option_selector_or_block_err.vv
+++ b/vlib/v/checker/tests/option_selector_or_block_err.vv
@@ -1,0 +1,17 @@
+struct Post {
+mut:
+  authors ?[]string
+}
+
+fn populate(mut post Post) []string {
+  return post.authors or {
+    post.authors = []string{}
+    post.authors
+  }
+}
+
+fn main() {
+  mut test := Post {}
+  populate(mut test)
+  println(test)
+}

--- a/vlib/v/checker/tests/wrong_none_on_or_block_err.out
+++ b/vlib/v/checker/tests/wrong_none_on_or_block_err.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/wrong_none_on_or_block_err.vv:8:22: error: wrong return type `none` in the `or {}` block, expected `int`
+vlib/v/checker/tests/wrong_none_on_or_block_err.vv:8:22: error: `or` block must provide a value of type `int`, not `none`
     6 |         `ő`: 25
     7 |     }
     8 |     println(m[`4`] or { none })

